File: lib/crc.ex
defmodule CRC do
  @moduledoc """
  This module is used to calculate CRC (Cyclic Redundancy Check) values
  for binary data. It uses NIF functions written in C to iterate over
  the given binary calculating the CRC checksum value.

  `CRC.Pure` provides the same API with no NIF, and will replace the NIF in
  v1.0. See the "Upgrading to 1.0" section of the README.

  CRC implementations have been tested against these online calculators to
  validate their correctness to the best of our ability.

  https://www.lammertbies.nl/comm/info/crc-calculation.html
  http://www.sunshine2k.de/coding/javascript/crc/crc_js.html
  """

  @doc """
  Calculate a CRC checksum for the `input` based on the crc `params` given.

  `params` can be an atom for one of the compiled models. See `CRC.list/0` for
  a full list or a Map with parameters to create a model at runtime. The map
  given should have all of the following keys:

  `width` - (unsigned integer) representation for the width of the CRC in bits
  `poly` - (unsigned integer) the polynomial used for the CRC calculation
  `init` - (unsigned integer) The initial value used when starting the calculation
  `refin` - (boolean) if the input value should be reflected. This is used for changing between endian's
  `refout` - (boolean) if the outvalue should be reflected when calculation is completed
  `xorout` - (unsigned integer) Final xor value used when completing the CRC calculation

  ## Examples

      %{
        width: 16,
        poly: 0x1021,
        init: 0x00,
        refin: false,
        refout: false,
        xorout: 0x00
      }

  You can also extend one of the compiled models at runtime by creating a map
  with `extend` key set to the model you wish to extend and the keys you wish
  to override for that model.

  For example to override the initial value for the `:crc_16_ccitt_false` model
  to `0x1D0F` you would pass the following Map as params:

      `%{extend: :crc_16_ccitt_false, init: 0x1D0F}`

  The input comes first so calls can be written as pipelines:

      read_data() |> CRC.calculate(:crc_16) |> do_something()

  You can learn more about CRC calculation here:
  https://www.sunshine2k.de/articles/coding/crc/understanding_crc.html
  """
  @spec calculate(iodata(), :crc_algorithm.params()) :: :crc_algorithm.value()
  defdelegate calculate(input, params), to: :crc

  @doc """
  Initialize a resource to be used for a multi-part CRC calculation with
  `update/2` and `final/1`.

  Resource is created using the same `params` types that are used with
  `calculate/2`:

    - atom's for compiled models
    - Map with model values
    - Map to extend a compiled model.
  """
  @spec init(:crc_algorithm.params()) :: :crc_algorithm.resource()
  defdelegate init(params), to: :crc

  @doc """
  Begins or continues a multi-part CRC calculation.

  Takes a `resource` from result of `init/1` or previous `update/2`
  call, and binary `input`, returns a new `resource` to be used to continue or
  finalize the CRC calculation.
  """
  @spec update(:crc_algorithm.resource(), iodata()) :: :crc_algorithm.resource()
  defdelegate update(resource, input), to: :crc

  @doc """
  Takes a `resource` result from `update/2` and finalizes the multi-part
  CRC calculation.
  """
  @spec final(:crc_algorithm.resource()) :: :crc_algorithm.value()
  defdelegate final(resource), to: :crc

  @doc """
  Returns the parameters (`width`, `poly`, `init`, `refin`, `refout`,
  `xorout`, `check`, `residue` and `sick`) of a model or of a resource
  returned by `init/1`.
  """
  @spec info(:crc_algorithm.params() | :crc_algorithm.resource()) :: :crc_algorithm.info()
  defdelegate info(params_or_resource), to: :crc

  @doc """
  Returns the residue of a model or of a resource returned by `init/1`.
  """
  @spec residue(:crc_algorithm.params() | :crc_algorithm.resource()) :: :crc_algorithm.value()
  defdelegate residue(params_or_resource), to: :crc

  @doc """
  Returns a list of all the compiled CRC models.
  """
  @spec list() :: [{atom, String.t()}]
  defdelegate list(), to: :crc

  @doc """
  Returns a list of all compiled CRC Models that match the filter given.

  Filter is compiled into a regular expression and matched against the model name
  and description.
  """
  @spec list(binary) :: [{atom, String.t()}]
  defdelegate list(filter), to: :crc

  @doc """
  Calculates an XOR checksum for the given binary
  """
  @spec checksum_xor(binary) :: number
  defdelegate checksum_xor(input), to: :crc

  @doc "Deprecated, use `CRC.calculate/2` instead (note the argument order: input, params)."
  @deprecated "Use CRC.calculate/2 instead (note the argument order: input, params)"
  @spec crc(:crc_algorithm.params(), iodata()) :: :crc_algorithm.value()
  def crc(params, input), do: :crc_fast.calc(params, input)

  @doc "Deprecated, use `CRC.init/1` instead."
  @deprecated "Use CRC.init/1 instead"
  @spec crc_init(:crc_algorithm.params()) :: :crc_algorithm.resource()
  def crc_init(params), do: :crc_fast.init(params)

  @doc "Deprecated, use `CRC.update/2` instead."
  @deprecated "Use CRC.update/2 instead"
  @spec crc_update(:crc_algorithm.resource(), iodata()) :: :crc_algorithm.resource()
  def crc_update(resource, input), do: :crc_fast.update(resource, input)

  @doc "Deprecated, use `CRC.final/1` instead."
  @deprecated "Use CRC.final/1 instead"
  @spec crc_final(:crc_algorithm.resource()) :: :crc_algorithm.value()
  def crc_final(resource), do: :crc_fast.final(resource)

  use CRC.Legacy
end

File: src/crc.erl
%% -*- mode: erlang; tab-width: 4; indent-tabs-mode: 1; st-rulers: [70] -*-
%% vim: ts=4 sw=4 ft=erlang noet
-module(crc).

%% Public API
-export([calculate/2]).
-export([init/1]).
-export([update/2]).
-export([final/1]).
-export([info/1]).
-export([residue/1]).
-export([list/0]).
-export([list/1]).
-export([checksum_xor/1]).
%% Deprecated API, removed in v1.0
-export([crc/2]).
-export([crc_init/1]).
-export([crc_update/2]).
-export([crc_final/1]).
-export([crc_8/1]).
-export([crc_8/2]).
-export([crc_16/1]).
-export([ccitt_16/1]).
-export([ccitt_16/2]).
-export([ccitt_16_kermit/1]).
-export([ccitt_16_kermit/2]).
-export([ccitt_16_xmodem/1]).
-export([ccitt_16_1D0F/1]).
-export([crc_16_dnp/1]).
-export([crc_16_modbus/1]).
-export([crc_16_sick/1]).
-export([crc_32/1]).
%% Internal API
-export([priv_dir/0]).

-deprecated([
	{crc, 2, "use crc:calculate/2 instead (note the argument order: Input, Model)"},
	{crc_init, 1, "use crc:init/1 instead"},
	{crc_update, 2, "use crc:update/2 instead"},
	{crc_final, 1, "use crc:final/1 instead"},
	{crc_8, 1, "use crc:calculate(Input, #{extend => crc_8_koop, init => 0}) instead"},
	{crc_8, 2, "use crc:calculate(Input, #{extend => crc_8_koop, init => Init}) instead, "
		"where Init is the bit-reflected value of (Seed bxor 16#FF)"},
	{crc_16, 1, "use crc:calculate(Input, crc_16) instead"},
	{ccitt_16, 1, "use crc:calculate(Input, crc_16_ccitt_false) instead"},
	{ccitt_16, 2, "use crc:calculate(Input, #{extend => crc_16_ccitt_false, init => Seed}) instead"},
	{ccitt_16_kermit, 1, "use crc:calculate(Input, crc_16_kermit) instead"},
	{ccitt_16_kermit, 2, "use crc:calculate(Input, #{extend => crc_16_kermit, init => Seed}) instead"},
	{ccitt_16_xmodem, 1, "use crc:calculate(Input, xmodem) instead"},
	{ccitt_16_1D0F, 1, "use crc:calculate(Input, #{extend => crc_16_ccitt_false, init => 16#1D0F}) instead"},
	{crc_16_dnp, 1, "use crc:calculate(Input, crc_16_dnp) instead"},
	{crc_16_modbus, 1, "use crc:calculate(Input, crc_16_modbus) instead"},
	{crc_16_sick, 1, "use crc:calculate(Input, crc_16_sick) instead"},
	{crc_32, 1, "use crc:calculate(Input, crc_32) instead"}
]).

%% Types
-type uint8_t() :: 16#00..16#FF.
-type uint16_t() :: 16#0000..16#FFFF.
-type uint32_t() :: 16#00000000..16#FFFFFFFF.

%%%===================================================================
%%% Public API Functions
%%%===================================================================

%% @doc Calculates the CRC of `Input' using `Model': a pre-defined model
%% name or alias (see list/0), a map of model parameters, or a map with an
%% `extend' key naming a pre-defined model plus the parameters to override.
-spec calculate(iodata(), crc_algorithm:params()) -> crc_algorithm:value().
calculate(Input, Model) ->
	crc_fast:calc(Model, Input).

%% @doc Starts a multi-part CRC calculation for `Model'.
-spec init(crc_algorithm:params()) -> crc_algorithm:resource().
init(Model) ->
	crc_fast:init(Model).

%% @doc Continues a multi-part CRC calculation with `Input', returning a
%% new resource to pass to the next update/2 or final/1 call.
-spec update(crc_algorithm:resource(), iodata()) -> crc_algorithm:resource().
update(Resource, Input) ->
	crc_fast:update(Resource, Input).

%% @doc Finishes a multi-part CRC calculation and returns the CRC.
-spec final(crc_algorithm:resource()) -> crc_algorithm:value().
final(Resource) ->
	crc_fast:final(Resource).

%% @doc Returns the parameters of a model or of a resource from init/1.
-spec info(crc_algorithm:params() | crc_algorithm:resource()) -> crc_algorithm:info().
info(ModelOrResource) ->
	Info = crc_fast:info(to_resource(ModelOrResource)),
	maps:with([width, poly, init, refin, refout, xorout, check, residue, sick], Info).

%% @doc Returns the residue of a model or of a resource from init/1.
-spec residue(crc_algorithm:params() | crc_algorithm:resource()) -> crc_algorithm:value().
residue(ModelOrResource) ->
	crc_fast:residue(to_resource(ModelOrResource)).

%% @doc Returns every pre-defined model as `{Key, Name}'.
-spec list() -> [{atom(), binary()}].
list() ->
	crc_models:list().

%% @doc Returns the pre-defined models whose key or name matches the
%% regular expression `Filter'.
-spec list(iodata()) -> [{atom(), binary()}].
list(Filter) ->
	case re:compile(Filter) of
		{ok, Regex} ->
			[Model || Model = {Key, Name} <- list(),
				re:run(atom_to_binary(Key, utf8), Regex, [{capture, none}]) =:= match
				orelse re:run(Name, Regex, [{capture, none}]) =:= match];
		{error, _} ->
			erlang:error({badarg, [Filter]})
	end.

%% @doc Calculates an 8-bit XOR checksum of `Input'.
-spec checksum_xor(binary()) -> uint8_t().
checksum_xor(Input) ->
	crc_nif:checksum_xor(Input).

%%%===================================================================
%%% Deprecated API Functions
%%%===================================================================

-spec crc(crc_algorithm:model(), iodata()) -> crc_algorithm:value().
crc(Model, Input) ->
	crc_fast:calc(Model, Input).

-spec crc_init(crc_algorithm:model()) -> crc_algorithm:resource().
crc_init(Model) ->
	crc_fast:init(Model).

-spec crc_update(crc_algorithm:resource(), iodata()) -> crc_algorithm:resource().
crc_update(Resource, Input) ->
	crc_fast:update(Resource, Input).

-spec crc_final(crc_algorithm:resource()) -> crc_algorithm:value().
crc_final(Resource) ->
	crc_fast:final(Resource).

-spec crc_8(binary()) -> uint8_t().
crc_8(Input) ->
	crc_8(Input, 16#FF).

-spec crc_8(binary(), uint8_t()) -> uint8_t().
crc_8(Input, Seed) ->
	crc_nif:crc_8(Seed, Input).

-spec crc_16(binary()) -> uint16_t().
crc_16(Input) ->
	crc_fast:calc(crc_16, Input).

-spec ccitt_16(binary()) -> uint16_t().
ccitt_16(Input) ->
	crc_fast:calc(crc_16_ccitt_false, Input).

-spec ccitt_16(binary(), uint16_t()) -> uint16_t().
ccitt_16(Input, Seed) ->
	crc_fast:calc(#{extend => crc_16_ccitt_false, init => Seed}, Input).

-spec ccitt_16_kermit(binary()) -> uint16_t().
ccitt_16_kermit(Input) ->
	crc_fast:calc(crc_16_kermit, Input).

-spec ccitt_16_kermit(binary(), uint16_t()) -> uint16_t().
ccitt_16_kermit(Input, Seed) ->
	crc_fast:calc(#{extend => crc_16_kermit, init => Seed}, Input).

-spec ccitt_16_xmodem(binary()) -> uint16_t().
ccitt_16_xmodem(Input) ->
	crc_fast:calc(xmodem, Input).

-spec ccitt_16_1D0F(binary()) -> uint16_t().
ccitt_16_1D0F(Input) ->
	crc_fast:calc(#{extend => crc_16_ccitt_false, init => 16#1D0F}, Input).

-spec crc_16_dnp(binary()) -> uint16_t().
crc_16_dnp(Input) ->
	crc_fast:calc(crc_16_dnp, Input).

-spec crc_16_modbus(binary()) -> uint16_t().
crc_16_modbus(Input) ->
	crc_fast:calc(crc_16_modbus, Input).

-spec crc_16_sick(binary()) -> uint16_t().
crc_16_sick(Input) ->
	crc_fast:calc(crc_16_sick, Input).

-spec crc_32(binary()) -> uint32_t().
crc_32(Input) ->
	crc_fast:calc(crc_32, Input).

%%%===================================================================
%%% Internal API Functions
%%%===================================================================

-spec priv_dir() -> file:filename_all().
priv_dir() ->
	case code:priv_dir(?MODULE) of
		{error, bad_name} ->
			case code:which(?MODULE) of
				Filename when is_list(Filename) ->
					filename:join([filename:dirname(Filename), "../priv"]);
				_ ->
					"../priv"
			end;
		Dir ->
			Dir
	end.

%%%-------------------------------------------------------------------
%%% Internal functions
%%%-------------------------------------------------------------------

%% @private
to_resource(Resource) when is_reference(Resource) ->
	Resource;
to_resource(Model) ->
	crc_fast:init(Model).
